"Sign in With Apple" provider

martincostello opened the pull request:
Adds a _Sign In with Apple_ provider based on currently available information.

Reference sources were:
  * [Aaron Parecki's _"What the Heck is Sign In with Apple?"_](https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple)
  * [Bruno Kreb's _"What is Sign In with Apple: Learn About the New Identity Provider"_](https://auth0.com/blog/what-is-sign-in-with-apple-a-new-identity-provider/)
  * [Apple's _Sign In with Apple REST API_ documentation](https://developer.apple.com/documentation/signinwithapplerestapi)

Relates to #314.

All the provider does at present is set the `NameIdentifier` claim in the claims principal to the value of the subject of the JWT returned from the token endpoint.

No information about a user information endpoint is yet available, but the [docs](https://developer.apple.com/documentation/signinwithapplerestapi/tokenresponse) suggest this will come at a later point:

> **`access_token`** `string`
> (Reserved for future use) A token used to access allowed data. Currently, no data set has been defined for access.

martincostello committed:
Basic Apple provider

Add a very basic Sign In with Apple provider based on currently available information.

kinosang commented on src/AspNet.Security.OAuth.Apple/AppleAuthenticationOptions.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    /// Defines a set of options used by <see cref="AppleAuthenticationHandler"/>.
+    /// </summary>
+    public class AppleAuthenticationOptions : OAuthOptions
```
According to the [docs](https://developer.apple.com/documentation/signinwithapplerestapi/generate_and_validate_tokens#3262048), I think it should have an option field for a private key, so that the provider can generate the credentials programmatically.

martincostello replied:
Yeah, I thought about that, but I’ll need to sort out a working setup in Apple first so I’ve got something to play around with to test it with 🙂

martincostello replied:
This is now possible by setting `GenerateClientSecret` to `true` and then either returning the private key's bytes via the `Func<string, Task<byte[]>> PrivateKeyBytes` property or providing a custom implementation of the `AppleKeyStore` class.

martincostello committed:
Implement Apple provider

Fully implement the provider for Sign In with Apple based on current available functionality.

martincostello committed:
Enable Sign In with Apple

Enable Sign In with Apple in the MVC sample app without hard-coding secrets.

martincostello committed:
Update tests

Update the tests for the updated provider implementation.
Fix incorrect test method name.

martincostello committed:
Enable token lifetime validation

Enable the validation of token lifetimes.

martincostello commented:
Working demo now available at https://signinwithapple.azurewebsites.net/

martincostello committed:
Add null annotations

Add [NotNull] attributes to relevant methods.

martincostello committed:
Improve exception handling

Pre-validate the ID token has a value.
Change catch clause to improve logging.

martincostello committed:
Extend integration tests

Extend the integration tests for additional scenarios such as no validation, invalid tokens and using a configured client secret.

martincostello committed:
Move expiry period to options

Move the configured lifetime for generated client secrets to the options class.

martincostello committed:
Add ClientSecretExpiresAfter validation

Add validation for the value of ClientSecretExpiresAfter.

martincostello committed:
Add tests for options validation

Add unit tests for options validation.
Improve exception type.
Fix incorrect if condition that meant not all values were validated correctly.

martincostello committed:
Add unit tests for client secret

Add unit tests for the generated client secret's format.

martincostello committed:
Make KeyId required

Make the KeyId option required if GenerateClientSecret is true.

martincostello committed:
Fix test

Fix the expiry not being set.

martincostello commented:
Note that in Azure App Service, you have to set `WEBSITE_LOAD_USER_PROFILE=1` to be able to load the private key from the `.p8` file.

martincostello commented:
Works on Windows but fails on macOS and Linux. Needs refactoring to account for [`CngKey.Import()`](https://github.com/martincostello/AspNet.Security.OAuth.Providers/blob/2a0bc93e6e7b4e5f1d6d0b13e0fe565ed1b243a5/src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleClientSecretGenerator.cs#L82) being Windows-specific: https://github.com/dotnet/corefx/issues/18733#issuecomment-296723615.

martincostello committed:
Fix Linux and macOS secret generation

Work around platform differences between Windows and Linux/macOS by supporting .p12/.pfx certificates for Linux/macOS and using p8 for Windows.
.NET Core 3.0 adds support for .p8 on both platforms.

martincostello committed:
Add password option for pfx files

Add an option for specifying a password for PFX files.
Add a test private key that has a password for use on macOS.

kinosang commented:
Have you try with `X509Certificate2`? It can parse pkcs8 or pkcs12, and provide `RsaPrivateKey` and `RsaPublicKey`.

martincostello committed:
Fix flaky test

Fix flaky test by setting the expiry to 2 seconds to eliminate rounding issues.

martincostello commented:
Yeah, I found a comment that lead me to `X509Certificate2` for if the private key is in `.pfx` format.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/a088f736c2ebc2314f93195cd61f14aadb946573/src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleClientSecretGenerator.cs#L96-L121

For now I've forked the implementation so Windows supports `.p8` (as that's the format Apple provide the user with so is the least-work to integrate), but Linux and macOS only support `.pfx` so the user needs to extract the `.cer` from the `.p8` and then combine them into a `.pfx` using OpenSSL.

Now the tests pass: https://travis-ci.com/martincostello/AspNet.Security.OAuth.Providers/builds/114862873

.NET Core 3.0 supports `.p8` on all three platforms, so the implementation can be simplified for 3.0 and the #280 branch, but I imagine if the community want to use this when Apple release the product, they'd rather have the option of 2.x or 3.0, rather than having to port their app to 3.0 just to be able to use _Sign In with Apple_.

martincostello commented:
ASP.NET Core 3.0 branch of the Apple provider is [here](https://github.com/martincostello/AspNet.Security.OAuth.Providers/tree/AspNetCore-300-Apple).

The code is simpler and supports `.p8` private keys on all platforms: ee4a2f7ca3d0f1550d3f3990a63e9fb01158caf9

Tests: https://travis-ci.com/martincostello/AspNet.Security.OAuth.Providers/builds/114865903

martincostello committed:
Add UsePrivateKey() method

Add UsePrivateKey() extension method that configures a private key file to use to auto-generate client secrets.

martincostello committed:
Bump System.IdentityModel.Tokens.Jwt

Bump System.IdentityModel.Tokens.Jwt to 5.3.0 to ensure that incompatibility with .NET Standard 1.4 doesn't affect consumers.

martincostello commented:
I've done a blog post about the implementation here: https://blog.martincostello.com/sign-in-with-apple-prototype-for-aspnet-core/

epoyraz commented:
@martincostello any update on this issue?

martincostello commented:
@epoyraz None as of yet, unless Apple have updated anything since the initial announcement?

epoyraz commented:
@martincostello ah, i see. Thanks for the info! 👍

dylanbevandotnet commented:
Hey @martincostello I'm not sure on how the non-windows flow works with the pfx. It looks like setting the key name to a key.pfx instead of key.p8 will still attempt to read the cert content as a base64 string which fails. Could you please give some guidance on how it's intended to work? If I manually read the cert via  509Certificate2 passing in the password and using that rawBytes value as the base64 string I get a crypto failure later on

martincostello commented:
Hi @dylanbevandotnet - does the _Cross-platform Support_ section of [my blog post](https://blog.martincostello.com/sign-in-with-apple-prototype-for-aspnet-core/) explain things? 

Essentially, the underlying cryptographic APIs in .NET Core 2.2 behave differently, so there's a different approach per-platform.

This won't be a problem in 3.0 as the APIs have feature parity in this area.

dylanbevandotnet commented:
@martincostello not really I'm afraid. I don't see how a password protected pfx is read in as a base64 string. Following your blog I would expect that I replace the UsePrivateKey delegate to point to the pfx instead of the p8, and supply the certificate password. However doing that will cause the exception to be thrush in the extension method as it tries to convert that pfx content to a base 64 string.

martincostello commented:
Is it possible you've generated the pfx file in an incorrect format?

Does your pfx look like a similar file format to the one in the tests? https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/99dadcdba64b3a8b8ed1e477c54e052abfb68e9f/test/AspNet.Security.OAuth.Providers.Tests/Apple/test.pfx

dylanbevandotnet commented:
@martincostello so the root issue is that if you use a non Windows os you have to set the `PrivateKeyBytes` delegate because if you just set the `UsePrivateKey` delegate it'll automatically assume that you're not using a pfx because it contains a call to `ConvertFromBase64String`. If the code in `UsePrivateKey` were to catch a `System.FormatException` it could then fall back to just returning the raw bytes from the file. From an end user's view point it seems nicer that I can point to either a p8 or a pfx and the code will figure out the right thing to do.

martincostello commented:
Catching an exception at runtime seems a bit bleurgh, and ultimately is a hack to workaround the different API surfaces.

For 2.2, the caller is the owner of the key material and knows their own runtime environment, so should make the appropriate decision on how to configure things.

There's an equally-hacky way to do this here: https://github.com/martincostello/SignInWithAppleSample/blob/50d0c6eb04dca53d0cd69f16d3005bf631291af0/src/SignInWithApple/Apple/AppleAuthenticationOptionsExtensions.cs#L39-L57

For 3.0 the need to do that becomes redundant as all platforms can natively use the `.p8` file Apple provide with no extra complexity. See here for the changes: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/ee4a2f7ca3d0f1550d3f3990a63e9fb01158caf9

That said, I can always re-review things once I do further work on this whenever Apple provide the capability to get the user claims for the token and this code can progress further towards being feature-complete.

dylanbevandotnet commented:
Sounds reasonable. Thanks @martincostello

kevinchalet commented:
@martincostello is there any x-plat issue preventing you from using `X509Certificate2` everywhere? 

Importing the private key bytes certainly works, but you won't be able to use HSM-backed EC certificates using this approach.

martincostello commented:
@PinpointTownes There was an issue with using a pfx cert with no password on it (admittedly, not a secure practice, my test cert found it), so it wasn't 100% compatible. That's what lead to me forking the setup code for 2.2 after I hit a brick wall trying to get `.p8` to work in the first place.

The 3.0 version can be nicer as it can use the `.p8` file Apple provide from the developer portal that doesn't have a password set on it on Windows, macOS and Linux, so is simpler to configure.

Relevant excerpt from my blog post:

> Once the Client Secret was being generated successfully from the .p8 file as-needed, I figured the bulk of the implementation was pretty much done.

> As it turned out, I'd made some assumptions in the implementation from working on my Windows 10 laptop that meant that the provider only worked on Windows and not on Linux or macOS, which is a bit embarrassing for integrating with an Apple product, let alone on cross-platform .NET Core.

> First this required changing the code so it could use more generic ECDA APIs to load the .p8 private key from Linux and macOS. A bit of Google-fu later lead to me finding this issue in Core CLR. It turns out that PKCS #8 keys aren't supported in .NET Core 2.x on non-Windows platforms. This is fixed in .NET Core 3.0, but that doesn't do us any good right now.

> To fix this I followed these instructions to generate a .pfx file (PKCS #12) from the .p8 file and then use the X509Certificate2 class to load the key instead (commit) on non-Windows platforms. It's a bit bleurgh to have to branch the code based on the operating system and require a different key format, but without pulling in a lot of extra code I didn't think it was worth it.

> Unfortunately, that didn't fix everything either. It fixed Linux but macOS was still broken. This time it was because of this issue where macOS cannot open private keys with no password set.

This required me to add a further option to support specifying a password for the certificate, which on reflection I should have done anyway. I was just being lazy in my tests.

With that change done, finally everything was working as expected on both Windows, Linux and macOS!

dylanbevandotnet commented:
Looks like Apple have changed something. I now see `response_mode must be form_post when name or email scope is requested.`

martincostello commented:
Yeah, my test site has stopped working too. I’ll have to investigate when I get the time over the weekend.

martincostello commented:
So the `response_mode` error is resolved by adding this to `AppleAuthenticationHandler`, but then it fails after sign-in with: _"The oauth state was missing or invalid."_.

```csharp
protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
{
    string challengeUrl = base.BuildChallengeUrl(properties, redirectUri);
    return QueryHelpers.AddQueryString(challengeUrl, "response_mode", "form_post");
}
```

kevinchalet commented:
@martincostello the OAuth 2.0 generic handler doesn't support `response_mode=form_post` (only the implicit `response_mode=query`).

@Tratcher there's no plan to update it to support it, right?

martincostello commented:
The following _very-much-a-hack_ minimal solution gets things working again in my sample site: https://github.com/martincostello/SignInWithAppleSample/commit/cc2ec90a15f5b77d5280dcc9b51bab419e4dd47f

I guess a proper solution would be to take the code from the base-class as-is, and refactor it to use the parameters from a dictionary populated from either `Form` or `Query`, depending on the mode.

martincostello committed:
Set response_mode to form_post

React to changes Apple have made to the sign-in service, and use form_post as the response mode.
This requires reimplementing HandleRemoteAuthenticateAsync() by using either for Form or Query based on whether it is an HTTP POST.

Tratcher commented:
> @Tratcher there's no plan to update it to support it, right?

There's no open or closed issue tracking it, so no. I've only seen that used in the context of OpenIdConnect.

dylanbevandotnet commented:
@martincostello does this allow you to get the email back in the exchange? I've struggled to get access to that in the openid connect implementation I tried

martincostello commented:
@dylanbevandotnet No, not yet. Apple haven’t documented the service that you can use to exchange the access token for the name or email yet, so at the moment all you get is an identifier.

dylanbevandotnet commented:
Digging through the forms it seemed they used to just return it with the initial token, but now they claim it's returned with each back channel exchange.

________________________________
From: Martin Costello <notifications@github.com>
Sent: Sunday, September 8, 2019 9:28:29 AM
To: aspnet-contrib/AspNet.Security.OAuth.Providers <AspNet.Security.OAuth.Providers@noreply.github.com>
Cc: dylanbevandotnet <chimpfunk@live.ca>; Mention <mention@noreply.github.com>
Subject: Re: [aspnet-contrib/AspNet.Security.OAuth.Providers] [WIP] "Sign in With Apple" provider (#318)


@dylanbevandotnet<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdylanbevandotnet&data=02%7C01%7C%7C5ec323093fd54cf9456908d734799583%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637035569105186207&sdata=WvS0mp5v7BVCPMxJz49rHnshN2R1vWPMv3jXq6iq8OU%3D&reserved=0> No, not yet. Apple haven’t documented the service that you can use to exchange the access token for the name or email yet, so at the moment all you get is an identifier.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Faspnet-contrib%2FAspNet.Security.OAuth.Providers%2Fpull%2F318%3Femail_source%3Dnotifications%26email_token%3DAEVMBJ74OW45QXV64SD4HZLQIUR23A5CNFSM4HVKY4IKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6FTX4Q%23issuecomment-529218546&data=02%7C01%7C%7C5ec323093fd54cf9456908d734799583%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637035569105196264&sdata=HAUhVkCSbCqLy%2BPo3C79Brd7jDkAMn%2F6jdgs%2BfHEV%2BE%3D&reserved=0>, or mute the thread<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAEVMBJ2CMDU4F6VNZZTS25TQIUR23ANCNFSM4HVKY4IA&data=02%7C01%7C%7C5ec323093fd54cf9456908d734799583%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637035569105196264&sdata=N2FybdKqST%2BrdFe2eaVDDO8Of5YYJiHUiSUwZG2b7%2Bk%3D&reserved=0>.

martincostello commented:
@dylanbevandotnet You're right, I've dug through the forums a little more and _eventually_ worked out how to do it. Annoyingly you only get one shot at getting the name and email, which explains why I was getting confused not being able to find it.

I had to go to https://appleid.apple.com/account/manage for my account and click into this link and delete the association with my sample and then log in again to see the data.

![image](https://user-images.githubusercontent.com/1439341/64492381-2ecc8600-d26b-11e9-8162-0d231cfc4941.png)

I'll update the sample so it works with it, then I'll port it over to this PR.

martincostello committed:
Use latest C# version

Fix the build by enabling the latest version of C#.

martincostello committed:
Retrieve user details after sign-in

Get the user's name and email address, if available, as claims after signing in with an Apple ID. These details are only available the first time the user signs in; if they are not persisted they cannot currently be obtained again.

dylanbevandotnet commented:
Thanks @martincostello you're a star. This whole thing reminds me of the xkcd 'one more standard' comic

YuriiNskyi commented:
This pull request is great and completely fulfills requirements to use Apple Sign In, but are there any plans to finally merge it, due to upcoming Apple Event WWDC 19?

martincostello commented:
@YuriiNskyi There are no plans to merge this PR or release the provider until after Apple officially launch the service and the API surface is definitely stable.

YuriiNskyi commented:
@martincostello That's very disappointing fact, checking the last paragraph [in this news](https://developer.apple.com/news/?id=06032019j), we want to be prepared with already implemented Apple Sign In, when it goes to be commercially available.

martincostello commented:
@YuriiNskyi You're welcome to build the provider from source if you want to ship early, but this won't be merged until the Apple API surface is declared stable.

As you can see, I've already had to make breaking changes to the provider code over the weekend, and Apple's documentation is quite lacking and hasn't been updated since Sign In with Apple was first announced.

martincostello committed:
Update branding

Use "Sign in with Apple" instead of "Sign In with Apple".

martincostello committed:
Access events via options

Use the same approach as the other OAuth handlers and access the Events property via the Options property.

martincostello committed:
Resolve logging TODO

Remove TODO comment.
Check whether Trace logging is enabled before logging the Apple token response.

martincostello commented:
With iOS 13 scheduled for release next Thursday (the 19th September), I think this is basically "done" now and the service is probably unlikely to change further, at least as an initial release.

As this one's more involved that the other providers, would you mind giving this a review please @PinpointTownes if you get some time this coming week?

martincostello commented:
I’ll look into merging this either later today or over the weekend so it can also be incorporated into the 3.0.0 branch for Monday.

martincostello committed:
Comment out Apple option

Comment out the Apple provider as it causes the application to fail to start if the values aren't set and/or the key file does not exist.

martincostello commented:
A prerelease version of this package (`AspNet.Security.OAuth.Apple 2.1.1-preview-0882`) is available in [MyGet](https://myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.Apple).

martincostello commented:
There's also a version available for ASP.NET Core 3.0 RC1: `3.0.0-rc1.19470.58`

dylanbevandotnet commented:
It looks like things may have moved forward again.
https://bitbucket.org/openid/connect/src/default/How-Sign-in-with-Apple-differs-from-OpenID-Connect.md
________________________________
From: Martin Costello <notifications@github.com>
Sent: Friday, September 20, 2019 7:34:27 AM
To: aspnet-contrib/AspNet.Security.OAuth.Providers <AspNet.Security.OAuth.Providers@noreply.github.com>
Cc: dylanbevandotnet <chimpfunk@live.ca>; Mention <mention@noreply.github.com>
Subject: Re: [aspnet-contrib/AspNet.Security.OAuth.Providers] "Sign in With Apple" provider (#318)


There's also a version available for ASP.NET Core 3.0 RC1: 3.0.0-rc1.19470.58

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Faspnet-contrib%2FAspNet.Security.OAuth.Providers%2Fpull%2F318%3Femail_source%3Dnotifications%26email_token%3DAEVMBJ7NZUMINKX7YA3EUHTQKTNPHA5CNFSM4HVKY4IKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7G4JEQ%23issuecomment-533578898&data=02%7C01%7C%7C2ba265e7e3254a68ca5f08d73dd7a40e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637045868682596553&sdata=4ItyNscDSJBFsprKsF5wKTxOxXLktSX98K%2FEXBMeECY%3D&reserved=0>, or mute the thread<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAEVMBJ4SOG6W27HV77GDQK3QKTNPHANCNFSM4HVKY4IA&data=02%7C01%7C%7C2ba265e7e3254a68ca5f08d73dd7a40e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637045868682606562&sdata=fwm02ah7iXXRs%2BIbddM7RlYO5frbDHHfhDiT7dcCVII%3D&reserved=0>.

suencien commented:
Hi Martin, this is awesome! Will you be able to help? I'm using OpenId Connect at ASP.NET Web Forms to integrate with Sign in with Apple?

I've got it to authenticate, but when returning to the Return URI https://iluvrun.com/signin-apple, it hits the 404 error. How do I get the site to handle this?

martincostello commented:
> Hi Martin, this is awesome! Will you be able to help? I'm using OpenId Connect at ASP.NET Web Forms to integrate with Sign in with Apple?
> 
> I've got it to authenticate, but when returning to the Return URI https://iluvrun.com/signin-apple, it hits the 404 error. How do I get the site to handle this?

The provider is designed for ASP.NET Core, so I'm not sure how much success you'd get trying to get it to work as part of a ASP.NET Web Forms application.

The endpoints are provided as middleware, so I'm not sure how you'd go about trying to run that as part of a Web Forms app. I haven't worked with Web Forms since about 2013, so I don't really have the experience or bandwidth to help you out trying to get it to work with it either I'm afraid!

suencien commented:
Hi Martin, it's ok. It's been a great reference. I'll keep  searching for the answer :)
